Update turbulence models and add documentation (#932)
climate-machine/CLIMA: docs/src/Atmos/Model/turbulence.md

```@meta
EditURL = "<unknown>/src/Atmos/Model/turbulence.jl"
```

## Turbulence Closures
In `turbulence.jl` we specify turbulence closures. Currently,

> pointwise models of the eddy-viscosity eddy-diffusivity type are

Reviewer: ```suggestion
pointwise models of the eddy viscosity/eddy diffusivity type are
```

```

Fields

```@example turbulence
$(DocStringExtensions.FIELDS)
"""
struct ConstantViscosityWithDivergence{FT} <: TurbulenceClosure
    "Dynamic Viscosity [kg/m/s]"
    ρν::FT
end

vars_gradient(::ConstantViscosityWithDivergence, FT) = @vars()
vars_diffusive(::ConstantViscosityWithDivergence, FT) =
    @vars(S::SHermitianCompact{3, FT, 6})

function diffusive!(
    ::ConstantViscosityWithDivergence,
    ::Orientation,
    diffusive::Vars,
    ∇transform::Grad,
    state::Vars,
    aux::Vars,
    t::Real,
)

    diffusive.turbulence.S = symmetrize(∇transform.u)
end

function turbulence_tensors(
    m::ConstantViscosityWithDivergence,
    state::Vars,
    diffusive::Vars,
    aux::Vars,
    t::Real,
)

    S = diffusive.turbulence.S
    ν = m.ρν / state.ρ
    τ = (-2 * ν) * S + (2 * ν / 3) * tr(S) * I
    D_t = (ν isa Real ? ν : diag(ν)) * inv_Pr_turb
    return ν, D_t, τ
end
```

### [Smagorinsky-Lilly](@id smagorinsky-lilly)
The Smagorinsky turbulence model, with Lilly's correction to
stratified atmospheric flows, is included in CLIMA.
The input parameter to this model is the Smagorinsky coefficient.
For atmospheric flows, the coefficient `C_smag` typically takes values between

> 0.15 and 0.23. Flow dependent C_smag are currently not supported (e.g. Germano's

Reviewer: ```suggestion
0.15 and 0.23. Flow dependent `C_smag` are currently not supported (e.g. Germano's
```

extension). The Smagorinsky-Lilly model does not contain explicit filtered terms.
```math
\nu = (C_{s} \mathrm{f}_{b} \Delta)^2 \sqrt{|\mathrm{S}|}
```
with the stratification correction term
```math
\mathrm{f}_{b} = \sqrt{1 - \frac{\mathrm{Ri}}{\mathrm{Pr}_{t}}}
```\
$\mathrm{Ri}$ and $\mathrm{Pr}_{t}$ are the Richardson and
turbulent Prandtl numbers respectively.  $\Delta$ is the mixing length in the
relevant coordinate direction. We use the DG metric terms to determine the
local effective resolution (see `src/Mesh/Geometry.jl`), and modify the vertical lengthscale by the
stratification correction factor $\mathrm{f}_{b}$.

```@example turbulence
"""
    SmagorinskyLilly <: TurbulenceClosure

> See § 1.3.2 in CliMA documentation

Reviewer: ```suggestion
See CliMA-Atmos design document.
```

)
    transform.turbulence.θ_v = aux.moisture.θ_v
end

function diffusive!(
    ::SmagorinskyLilly,
    orientation::Orientation,
    diffusive::Vars,
    ∇transform::Grad,
    state::Vars,
    aux::Vars,
    t::Real,
)

    diffusive.turbulence.S = symmetrize(∇transform.u)
    ∇Φ = ∇gravitational_potential(orientation, aux)
    diffusive.turbulence.N² =
        dot(∇transform.turbulence.θ_v, ∇Φ) / aux.moisture.θ_v
end

function turbulence_tensors(
    m::SmagorinskyLilly,
    state::Vars,
    diffusive::Vars,
    aux::Vars,
    t::Real,
)

    FT = eltype(state)
    S = diffusive.turbulence.S
    normS = strain_rate_magnitude(S)
    k̂ = aux.orientation.∇Φ / grav
```

squared buoyancy correction

```@example turbulence
    Richardson = diffusive.turbulence.N² / (normS^2 + eps(normS))
    f_b² = sqrt(clamp(FT(1) - Richardson * inv_Pr_turb, FT(0), FT(1)))
    ν₀ = normS * (m.C_smag * aux.turbulence.Δ)^2 + FT(1e-5)
    ν = SVector{3, FT}(ν₀, ν₀, ν₀)
    ν_v = k̂ .* dot(ν, sqrt(f_b²) * k̂)
    ν_h = ν₀ .- ν_v
    τ = -2 * SDiagonal(ν_h .+ ν_v) * S
    D_t = (ν isa Real ? ν : diag(ν)) * inv_Pr_turb
    return SDiagonal(ν_h + ν_v), D_t, τ
end
```

### [Vreman Model](@id vreman)

> Vreman's turbulence model for anisotropic flows which provides a

Reviewer: ```suggestion
Vreman's turbulence model for anisotropic flows, which provides a
```

less dissipative solution (specifically in the near-wall and transitional regions)
than the Smagorinsky-Lilly method. This model

> relies of first derivates of the velocity vector (i.e. the gradient tensor).

Reviewer: ```suggestion
relies on first derivatives of the velocity vector (i.e., the gradient tensor).
```

By design, the Vreman model handles transitional as well as fully turbulent flows adequately.
The input parameter to this model is the Smagorinsky coefficient - the coefficient is modified
within the model functions to account for differences in model construction.
#### Equations
```math
\nu = 2.5 \mathrm{C}_{smag} \sqrt{\frac{\mathrm{B}_{\beta}}{\alpha_{i}\alpha_{j}}}
```
where $\mathrm{B}_{\beta}$ and $\alpha$ are functions of the velocity
gradient tensor terms.

```@example turbulence
"""
    Vreman{FT} <: TurbulenceClosure

>   §1.3.2 in CLIMA documentation

Reviewer: ```suggestion
  See CLIMA-Atmos design document.
```

    ∇transform::Grad,
    state::Vars,
    aux::Vars,
    t::Real,
)
    diffusive.turbulence.∇u = ∇transform.u
    ∇Φ = ∇gravitational_potential(orientation, aux)
    diffusive.turbulence.N² =
        dot(∇transform.turbulence.θ_v, ∇Φ) / aux.moisture.θ_v
end

function turbulence_tensors(
    m::Vreman,
    state::Vars,
    diffusive::Vars,
    aux::Vars,
    t::Real,
)
    FT = eltype(state)
    α = diffusive.turbulence.∇u
    S = symmetrize(α)
    k̂ = aux.orientation.∇Φ / grav

    normS = strain_rate_magnitude(S)
    Richardson = diffusive.turbulence.N² / (normS^2 + eps(normS))
    f_b² = sqrt(clamp(1 - Richardson * inv_Pr_turb, 0, 1))

    β = f_b² * (aux.turbulence.Δ)^2 * (α' * α)
    Bβ = principal_invariants(β)[2]

    ν₀ = m.C_smag^2 * FT(2.5) * sqrt(abs(Bβ / (norm2(α) + eps(FT))))

    ν₀ = SVector{3, FT}(ν₀, ν₀, ν₀)
    ν_h = cross(k̂, cross(ν₀, k̂))
    ν_v = k̂ .* dot(ν₀, sqrt(f_b²) * k̂)
    τ = -2 * SDiagonal(ν_h + ν_v) * S
    D_t = (ν isa Real ? ν : diag(ν)) * inv_Pr_turb
    return SDiagonal(ν_h + ν_v), D_t, τ
end
```

### [Anisotropic Minimum Dissipation](@id aniso-min-diss)
This method is based Vreugdenhil and Taylor's minimum-dissipation eddy-viscosity model.

> The principles of the Rayleigh-quotient minimiser are applied to the energy dissipation terms in the

Reviewer: ```suggestion
The principles of the Rayleigh quotient minimizer are applied to the energy dissipation terms in the
```

conservation equations, resulting in a maximum dissipation bound, and a model for
eddy viscosity and eddy diffusivity.
```math
\nu_e = (\mathrm{C}\delta)^2  \mathrm{max}\left[0, - \frac{\hat{\partial}_k \hat{u}_{i} \hat{\partial}_k \hat{u}_{j} \mathrm{\hat{S}}_{ij}}{\hat{\partial}_p \hat{u}_{q}} \right]
```

```@example turbulence
"""
    AnisoMinDiss{FT} <: TurbulenceClosure

>   §1.3.2 in CLIMA documentation

Reviewer: ```suggestion
```
